frame/collective/src/lib.rs

```diff
@@ -185,69 +329,185 @@ decl_module! {
 		/// - `new_members`: The new member list. Be nice to the chain and
 		//	provide it sorted.
 		/// - `prime`: The prime member whose vote sets the default.
+		/// - `old_count`: The number of previous members. (Used for weight estimation.)
 		///
 		/// Requires root origin.
-		#[weight = (100_000_000, DispatchClass::Operational)]
-		fn set_members(origin, new_members: Vec<T::AccountId>, prime: Option<T::AccountId>) {
+		///
+		/// # <weight>
+		/// ## Weight
+		/// - `O(MP + N)` where:
+		///   - `M` old-members-count (code- and governance-bounded)
+		///   - `N` new-members-count (code- and governance-bounded)
+		///   - `P` proposals-count (code-bounded)
+		/// - DB:
+		///   - 1 storage mutation (codec `O(M)` read, `O(N)` write) for reading and writing the members
+		///   - 1 storage read (codec `O(P)`) for reading the proposals
+		///   - `P` storage mutations (codec `O(M)`) for updating the votes for each proposal
+		///   - 1 storage write (codec `O(1)`) for deleting the old `prime` and setting the new one
+		/// - Benchmark: 0 + M * 20.47 + N * 0.109 + P * 26.29 µs (min squares analysis)
+		/// # </weight>
+		#[weight = (
+			weight_for::set_members(T::DbWeight::get(), *old_count, new_members.len() as Weight, T::MaxProposals::get()),
+			DispatchClass::Operational
+		)]
+		fn set_members(origin, new_members: Vec<T::AccountId>, prime: Option<T::AccountId>, old_count: u32) {
 			ensure_root(origin)?;
+			ensure!(new_members.len() <= T::MaxMembers::get() as usize, Error::<T, I>::TooManyMembers);
+			let old = Members::<T, I>::get();
+			ensure!(old.len() <= old_count as usize, Error::<T, I>::InvalidCount);
```

Comment:
Here we decode old members before checking the length.
Maybe it is ok if decoding old members doesn't cost that much. otherwise we can weight with T::MaxMembers and refund with the actual cost. (or do something similar to https://github.com/paritytech/substrate/pull/5828/files#diff-2e0725fa587f558456a6b465bbd8dc0bR1972 )

Reply:
We need to decode `old` anyway because it is passed into `set_members_sorted`.
But yes, max + refund might be better.

Reply:
(Yes I mean if the extrinsic fails at InvalidCount then the weight for the failing extrinsic can be to low compare to its cost.)

Reply:
It's unlikely to be too low, but yes in principle